posts/inside-rust/2020-06-24-lto-improvements.md

```diff
@@ -0,0 +1,89 @@
+---
+layout: post
+title: "Disk space and LTO improvements"
+author: Eric Huss
+description: "Disk space and LTO improvements"
+team: the Cargo team <https://www.rust-lang.org/governance/teams/dev-tools#cargo>
+---
+
+Thanks to the work of [Nicholas Nethercote] and [Alex Crichton], there have been some recent improvements that reduce the size of compiled libraries, and improves the compile-time performance, particularly when using LTO. This post dives into some of the details of what changed, and an estimation of the benefits.
+
+These changes have been added incrementally over the past three months, with the latest changes landing just a few days ago on the nightly channel. The bulk of the improvements will be found in the 1.46 stable release. It would be great for any projects that use LTO to test it out on the nightly channel and report any issues that arise.
+
+[Nicholas Nethercote]: https://github.com/nnethercote
+[Alex Crichton]: https://github.com/alexcrichton/
+
+## Background
+
+When compiling a library, `rustc` saves the output in an `rlib` file which is an [archive file]. This has historically contained these two things (among others):
+
+* Object code, which is the result of code generation. This is used during regular linking.
+* [LLVM bitcode], which is a binary representation of LLVM's intermediate representation. This can be used for [Link Time Optimization] (LTO).
+
+LTO is an optimization technique that can perform whole-program analysis. It analyzes all of the bitcode from every library at once, and performs optimizations and code generation. `rustc` supports several forms of LTO:
+
+* Fat LTO. This performs "full" LTO, which can take a long time to complete and may require a significant amount of memory.
+* [Thin LTO]. This is a lightweight version of "fat" LTO that can achieve similar performance improvements while taking much less time to complete.
```

Comment:
@alexcrichton  Because of the phrasing of this sentence, is there ever a time where fat LTO is better than thin LTO?

Reply:
Heh I wondered the same thing when ThinLTO came along! Fat LTO, however, produces smaller binaries because it can extremely aggressively internalize functions. It can also produce slightly faster binaries since total knowledge in a compiler is typically better than fragmented knowledge. ThinLTO's primary purpose is figure out what to inline across codegen units, while Fat LTO basically says "eh just do your normal thing but here's all the code in the world"

Reply:
That makes sense. Is there a way we can rephrase this to make this point more clear? The way it stands now makes it seem like thin is always better. "Similar performance while taking much less time"

Reply:
I've tweaked the wording a little to try to make it clearer.

I think as with any optimization settings, every project will need to experiment with different settings to see what works best for them.